dotcom-rendering/src/model/enhance-ad-placeholders.ts

```diff
@@ -31,28 +34,43 @@ const isSuitablePosition = (
 		return false;
 	}
 
-	// Insert an ad placeholder every `adEveryNParagraphs` paragraphs,
-	// starting from the paragraph at `firstAdIndex` and only after a paragraph
+	// Check that we haven't inserted an ad yet, and that we are far enough in to do so
+	const isFirstAdIndex = blockCounter >= firstAdIndex && lastAdIndex === 0;
+
+	// Check that we are at least `adEveryNParagraphs` blocks after the previous ad
+	const isEnoughBlocksAfter = blockCounter - lastAdIndex >= adEveryNBlocks;
+
+	// Insert an ad placeholder before the current element if it is a paragraph, the
+	// previous element was an image or a paragraph, and if the position is eligible
 	return (
 		isParagraph &&
-		(paragraphCounter - firstAdIndex) % adEveryNParagraphs === 0
+		prevIsParagraphOrImage &&
+		(isFirstAdIndex || isEnoughBlocksAfter)
 	);
 };
 
 const isParagraph = (element: FEElement) =>
 	element._type === 'model.dotcomrendering.pageElements.TextBlockElement';
 
-const insertPlaceholder = (elements: FEElement[]): FEElement[] => {
+const isImage = (element: FEElement) =>
+	element._type === 'model.dotcomrendering.pageElements.ImageBlockElement';
+
+const insertPlaceholder = (
+	prevElements: FEElement[],
+	currentElement: FEElement,
+): FEElement[] => {
 	const placeholder: AdPlaceholderBlockElement = {
 		_type: 'model.dotcomrendering.pageElements.AdPlaceholderBlockElement',
 	};
-	return [...elements, placeholder];
+	return [...prevElements, placeholder, currentElement];
 };
 
 type ReducerAccumulator = {
 	elements: FEElement[];
-	paragraphCounter: number;
+	blockCounter: number;
+	lastAdIndex: number;
```

Comment:
Is `lastAdIndex` the index of the most recent paragraph an ad has been inserted before? e.g. in the case that an ad has inserted before the 4th and 10th element, `lastAdIndex` would be 10? Maybe a comment would be useful here?

Reply:
Yes that's exactly it. We're using it to count that we have at least 6 blocks between ads. I can add a comment to make this clearer